feat!: API versioning and API v2 (beta)

ankush opened the pull request:
### Why?
- Versioning lets us improve the way APIs work while still being backward compatible. 
- We have known few problems with APIs for long time but they can't be easily addressed because they will likely require backward incompatible changes. 


### Changes

- All API routing is now handled by Werkzeug routing.
- New alias for API versions `/api/*` == `/api/v1/`
- `/api/v2` prefix for new APIs. 
- API versions will ONLY specify URL and response structure.
- REST APIs will not break between versions. RPC calls and schema are out of the scope of versioning, they CAN change between versions. Keeping them backward compatible is job of developers doing the changes.  



### V2 changes and features


##### 1. Shorthand for all whitelisted functions in controller files. 


```diff
- GET /api/v1/method/erpnext.selling.doctype.sales_order.sales_order.make_sales_invoice
+ GET /api/v2/method/Sales Order/make_sales_invoice
```

##### 2. Call whitelisted functions directly on ~~resources~~ => documents. 

```diff
- POST /api/v1/resource/Sales Order/SO-0001?run_method=submit
+ POST /api/v2/document/Sales Order/SO-0001/method/submit
```

closes https://github.com/frappe/frappe/issues/14869


##### 3. Response Structure change

- [x] `message` => `data` key for response data **everywhere**. 
- [x] `errors` key for nested and rich error information, replacing `exc`/`exc_type` etc. closes https://github.com/frappe/frappe/issues/14905


##### 4. URL/verb differences
- PATCH == PUT for updating documents. PATCH is the correct convention for partial updates.
- APIs are now divided in 3 types:
    - `document` - previously known as `resource`
    - `method` - RPC calls, same as before
    - `doctype` - new class of APIs that operate on doctype instead of documents. e.g. meta, count, list etc. 



### TODO

- [x] This PR currently copies a lot of code to demonstrate changes. Rewrite all code that touches v2 APIs before merge. 
- [x] Merge `handle.py` and `api.py`, remove `globals()` lookups in RPC handler. That's unnecessary special handling. 
- [x] URL encoding of names
- [x] Finalize response structure
- [x] Impl Errors and message response structure
- [x] Tests for both versions. Aim 100% coverage for this. 
- [x] Collections should have methods, e.g. `GET /api/resource/Sales Order/method/count`, meta (?)
- [x] File creation and attachment
- [ ] Update client side code, frappeclient.
    - [ ] xcall now uses v2
    - [x] All message handlers should work

Optional / separate PRs.
- [ ] Access to actual response object from whitelisted functions. 
- [ ] frappe.response.header for sending custom headers
- [ ] get_value alternate
- [ ] Documents should have methods like add attachment, get attachment, add comment, etc
- [ ] Bulk update
- [ ] permissions ??
- [ ] bulk insert with array
- [x] delete child doc
- [ ] update js `frappe.db`
- [ ] `?expand=["link_field_name"]` to link fields to their linked documents. 
- [ ] reports via rest api /api/report/name

ankush committed:
chore: deprecate form_dict.cmd, globals()

ankush committed:
refactor: Use werkzeug router for API routing

ankush committed:
refactor: create two API versions

ankush committed:
feat(APIv2): Execute doc method using API

e.g. `POST /api/v2/resource/Sales Order/SO-0001/submit`

ankush committed:
refactor!: Method whitelisting

Document.whitelist doesn't work, no idea why it's doing all weird
`__func__` business.

`@frappe.whitelist()` works just fine.

ankush committed:
feat: Shorthand for whitelisted function call in controllers

```diff
- GET /api/method/frappe.core.doctype.user.user.get_timezones
+ GET /api/method/User/get_timezones
```

ankush committed:
refactor!: move OAuth and token auth code to auth.py

This doesn't belong in api.py

ankush committed:
fix: rename type validation

`None` can be passed which is acceptable here.

TODO: Make slackdict accept none as bool and convert to False

ankush committed:
fix: double response processing

ankush committed:
fix: pass on custom responses as is without any post processing

ankush committed:
refactor(APIv2): `message` -> `data`

ankush committed:
refactor: remove unnessary manual commits

ankush committed:
test: check that document actually got deleted

ankush committed:
refactor: Drop `frappe.local` for proxies

ankush committed:
refactor: common API methods to utils

ankush committed:
refactor!: manual creation of response

This contains minor breaking change where `delete` API call now returns `OK` in `data` key instead of `message`.

ankush committed:
fix: delete oauth stuff when deleting users

ankush committed:
fix: route ambiguitities and `/` in name

v1 and v2:
- Use path convertor to consume entire document which MIGHT contain `/`
- Drop string convertor, as it's the default
- Dont merge slashes - avoid consuming something that's part of name

v2:
- `SI/BLAH/BLAH/submit` is not simple to parse without ambiguitity
- `SI/BLAH/BLAH/method/submit` removes ambiguitity for 99.99% cases.

netchampfaris commented:
- Collections should have methods, e.g. `GET /api/resource/Sales Order/method/count`, meta (?)
- Documents should have methods like add attachment, get attachment, add comment, etc
- get_value alternate
- bulk insert with array
- permissions ??
- File creation and attachment
- Bulk update
- frappe.response.header for sending custom headers
- Access to actual response object from whitelisted functions.

ankush committed:
chore: separate out v1 and v2

Any common utils end up becoming a problem, duplicating code here is
better than coupling potentially breaking behaviour

ankush committed:
refactor: simplify request form data

ankush committed:
test: refactor API test cases for paramterization

ankush committed:
test: basic tests v2 API

ankush committed:
refactor: OAuth flow without breaking routing convention

Appending `/connected-app` after method breaks routing.

ankush committed:
fix: backward compatiblity for method calls in v1

Preserves complete path just like before. This broke oauth as oauth code
was expecting 4th part in path even though it's never been supported in
API methods

ankush committed:
refactor: `resource` => `document`

This lets us create three types of APIs:

- Document APIs that operate on documents
- DocType APIs that operate on collections - list, count, search
- Method APIs that are RPC calls

ankush committed:
refactor!: merge handle.py

This has several breaking changes for v2:
1. No support for following methods which were implicitly present in
   default namespace.
    - run_doc_method
    - ping
    - web_logout
    - logout
    - uploadfile
    - upload_file
    - download_file

ankush committed:
fix!: Last overriden method should be considered

ankush committed:
feat: run_doc_method v2

ankush committed:
refactor: simplify v2 implementations

ankush committed:
fix: Better delete doc for child tables

gavindsouza commented on frappe/api/__init__.py:
```diff
@@ -0,0 +1,64 @@
+	- `/api/resource/{doctype}/{name}` will point to a resource
+	        `GET` will return document
+	        `POST` will insert
+	        `PUT` will update
```
`PATCH` could be implemented to update a few fields. Contrary to `PUT`, you can get away by passing a sparse Document?

gavindsouza replied:
Like `set_value` but with triggers? 🤔

gavindsouza replied:
OK, currently PUT == PATCH (update_doc implementation)

ankush replied:
Frappe's `PUT` IS `PATCH` if we go by conventions.

ankush replied:
https://github.com/frappe/frappe/pull/22300/commits/1834488f6174ad4f32df164822c85b2c20dfeb34

ankush replied:
We don't need PUT, idk any use cases so far.

ankush committed:
refactor: no need to cast bools manually

ankush committed:
fix: check for write permission if POST method

ankush committed:
fix: add login and logout methods

ankush committed:
refactor: PUT == PATCH

Correct conventions for partial updates

ankush committed:
feat: doctype collection level APIs - meta, count

ankush committed:
refactor: use reportview get_count implementation

Count reported should be according to permissions

ankush committed:
test: old API messages

ankush committed:
feat: v2 for message response strucutre

ankush committed:
refactor: defer unnecessary json-dumping of messages

Also avoid accessing locals where interface is present like for popping last message.

ankush committed:
feat: v2 error and debug log structure

ankush committed:
fix: Add old file upload for v2

netchampfaris commented on frappe/api/v2.py:
```diff
@@ -0,0 +1,193 @@
+	),
+	Rule("/method/<doctype>/<method>", endpoint=handle_rpc_call),
+	# Document level APIs
+	Rule("/document/<doctype>", methods=["GET"], endpoint=document_list),
+	Rule("/document/<doctype>", methods=["POST"], endpoint=create_doc),
```
Doctype list and create should be on `GET /doctype/ToDo` and `POST /doctype/ToDo` respectively?

`/document/ToDo/<name>/*` can be reserved for records that already exist?

ankush replied:
We can do both. I prefer POST in document instead of doctype personally. It's more REST-like.

ankush committed:
fix: assume v1 if path is not set

Old `cmd` calls will not work otherwise.

ankush committed:
feat: frappe.call can accept api_version

ankush committed:
test: Split API v2 tests

Duplication here seems better than weird and hard to understand
parameterized version.

ankush commented:
Merging this. We will do "stabilization" once we have a few internal users using this API version... till then this is _beta/internal/subject to change_ etc.
